_posts/2020-06-10-application-components.adoc

```diff
@@ -0,0 +1,574 @@
+
+=== Making the Petclinic Project an Application Component
+
+In order to leverage the Petclinic project as a product line, it is required to understand what makes up a CUBA application component and where it is different over a regular CUBA application. In fact, the difference is not very big - only one particular file called `app-component.xml`. This application component descriptor file marks a CUBA application additionally to be used as an application component. This means - it can be used by other CUBA applications.
+
+WARNING: In order to prevent a module naming collision between the application component and the application that is using it, it is required to change the module prefix of the Petclinic project from `app` to `petclinic` via the Project Properties window in CUBA Studio.
+
+Besides that, nothing really changed. The application component can still be used as a regular CUBA application. It can be started in CUBA Studio, business logic can be created, UIs can be generated etc.
+
+Used by other CUBA applications in this case means not only that the functionality of the application component shows up in the resulting application. Also, it is possible to interact with any part of the application component from inside the target application. Entities (like the `Pet` entity) can be referenced, Spring beans can be called or overridden, UI screens can be extended or replaced etc.
+
+Let's look into the `app-component.xml` of the Petclinic example:
+
+.app-component.xml
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<app-component xmlns="http://schemas.haulmont.com/cuba/app-component.xsd"
+  dependsOn="com.haulmont.cuba, com.haulmont.addon.helium">
+
+  <module name="global" blocks="*">
+    <artifact name="petclinic-global" appJar="true"/>
+
+    <property name="cuba.persistenceConfig" value="+com/haulmont/sample/petclinic/persistence.xml"/>
+    <property name="cuba.metadataConfig" value="+com/haulmont/sample/petclinic/metadata.xml"/>
+    <property name="cuba.viewsConfig" value="+com/haulmont/sample/petclinic/views.xml"/>
+  </module>
+  <module name="core" dependsOn="global" blocks="core">
+    <artifact name="petclinic-core" appJar="true"/>
+    <property name="cuba.mainMessagePack" value="+com.haulmont.sample.petclinic.core"/>
+    <!-- ... -->
+  </module>
+  <!-- ... -->
+</app-component>
+----
+
+This file contains a formal description of the structure of the application component. It exposes a couple of `module` elements. Those modules can declare `artifact` elements as well as `property` elements that should be exposed for this application component.
+
+CUBA mainly uses this information for configuring the right application properties (and values) as well as downloading the correct dependencies.
```

Comment:
Strictly speaking, downloading of dependencies is defined only by the build system, i.e. `build.gradle`.
 In app-component.xml we define what artifacts are packed in what app blocks (web, core).

Reply:
👍